core/src/main/java/com/linecorp/armeria/client/endpoint/ResourceFileEndpointGroup.java

```diff
@@ -0,0 +1,95 @@
+import java.util.Properties;
+
+import com.linecorp.armeria.client.Endpoint;
+
+/**
+ * A resource file backed {@link EndpointGroup}. The list of {@link Endpoint}s are loaded from the resource file
+ * in a current class loader.
+ */
+public final class ResourceFileEndpointGroup implements EndpointGroup {
+    /**
+     * Creates a new {@link EndpointGroup} instance that loads the host names (or IP address) and the port
+     * numbers of the {@link Endpoint} from the {@code resourceName} resource file. The resource file must
+     * contain at least one property whose name starts with {@code endpointKeyPrefix}:
+     *
+     * <pre>{@code
+     * example.hosts.0=example1.com:36462
+     * example.hosts.1=example2.com:36462
+     * example.hosts.2=example3.com:36462
+     * }</pre>
+     *
+     * @param resourceName the resource file name that the list of {@link Endpoint}s loaded from
+     * @throws IllegalArgumentException if failed to load any hosts from the specified resource file
+     */
+    public static ResourceFileEndpointGroup of(ClassLoader classLoader, String resourceName,
+                                               String endpointKeyPrefix, int defaultPort) {
+        return new ResourceFileEndpointGroup(loadEndpoints(classLoader, resourceName,
+                                                           endpointKeyPrefix, defaultPort));
+    }
+
+    private static List<Endpoint> loadEndpoints(ClassLoader classLoader, String resourceName,
+                                                String endpointKeyPrefix, int defaultPort) {
+        URL resourceUrl = classLoader.getResource(resourceName);
+        if (resourceUrl == null) {
+            throw new IllegalArgumentException(resourceName + " not found");
+        }
+        try (InputStream in = resourceUrl.openStream()) {
+            final List<Endpoint> newEndpoints = new ArrayList<>();
+            final Properties props = new Properties();
+            props.load(in);
+            for (Entry<Object, Object> e : props.entrySet()) {
+                final String key = (String) e.getKey();
+                final String value = (String) e.getValue();
+
+                if (key.startsWith(endpointKeyPrefix)) {
+                    final Endpoint endpoint = Endpoint.parse(value);
+                    checkState(!endpoint.isGroup(),
+                               "%s contains an endpoint group which is not allowed: %s", resourceName, value);
+                    newEndpoints.add(endpoint.withDefaultPort(defaultPort));
+                }
+            }
+            checkArgument(!newEndpoints.isEmpty(), "%s contains no hosts.", resourceName);
+            return newEndpoints;
+        } catch (IOException e) {
+            throw new IllegalArgumentException("failed to load: " + resourceName, e);
+        }
+    }
+
+    private final List<Endpoint> endpoints;
+
+    private ResourceFileEndpointGroup(List<Endpoint> endpoints) {
+        this.endpoints = endpoints;
```

Comment:
nit: How about making this list immutable?

Reply:
Changed `loadEndpoints` to return immutable list instead.